snowflake/custom-recipes/python-custom-recipes/recipe.py

```diff
@@ -1,113 +1,173 @@
             print("[-] Please check and correct your Project Variables.")
             sys.exit("Project Variables error")
     else:
         # Looking into Global Variables
+        print("[+] Could not find 'snowflake' in Project Variables so looking in Global Variables...")
         variables = dss.get_variables()
         if "snowflake" in variables:
             if "aws_access_key" in variables["snowflake"] and "aws_secret_key" in variables["snowflake"]:
                 print("[+] Found AWS credentials in Global Variables")
                 AWS_ACCESS_KEY = variables["snowflake"]["aws_access_key"]
                 AWS_SECRET_KEY = variables["snowflake"]["aws_secret_key"]
+            else:
+                print("[-] 'snowflake' key found in Global Variables but could not retrieve aws_access_key and/or aws_secret_key.")
+                print("[-] Please check and correct your Global Variables.")
+                sys.exit("Global Variables error")                
         else:
-            print("[-] Snowflake key found in Global Variables but can not retrieve aws_access_key and/or aws_secret_key.")
-            print("[-] Please check and correct your Global Variables.")
-            sys.exit("Global Variables error")
-    
+            print("[-] 'snowflake' key was not found in Project or Global Variables.")
+            print("[-] Please check and correct your Project or Global Variables.")
+            sys.exit("Project or Global Variables error")
+else:
+    print("[+] Use the S3 credentials specified in the Plugin interface.")
+    AWS_ACCESS_KEY = get_recipe_config().get("aws_access_key")
+    AWS_SECRET_KEY = get_recipe_config().get("aws_secret_key")
+    if not AWS_ACCESS_KEY or not AWS_SECRET_KEY:
+        print("[-] S3 credentials were not provided in the Plugin interface.")
+        sys.exit("Plugin Variables error")
+
 
-# Dataiku Datasets
-ds = dataiku.Dataset(DATASET_IN)
-out = dataiku.Dataset(DATASET_OUT)
 
 
 #------------------------------------------------------------------------------
 # INPUT DATASET SETTINGS
 #------------------------------------------------------------------------------
 
 # Input dataset settings
-config = ds.get_config()
 
-if config["formatType"] != 'csv':
-    print("[-] Only a CSV format for the input DSS Dataset is supported (you used {}).".format(config["formatType"]))
+if config_in["formatType"] != 'csv':
+    print("[-] Only a CSV format for the input DSS Dataset is supported (you used {}).".format(config_in["formatType"]))
     print("[-] Please adjust the format. Aborting")
     sys.exit("Format error (CSV needed)")
 
-project_key = config["projectKey"]
+project_key = config_in["projectKey"]
 
 # Actual path of the input file on S3
-bucket = config["params"]["bucket"]
-path = config["params"]["path"].replace("${projectKey}",config["projectKey"])
+bucket = config_in["params"]["bucket"]
```

Comment:
the "bucket" key will be accessible only if the dataset has been built by DSS. In the case of external dataset this will raise an error. For external datasets the information can be accessed in the "path" key of the dataset config. Better having the two options taken into account here (maybe a plugin python-lib with specific functions that read access this information so that it can be more easily unit testable)

Reply:
How do y'all run unit tests on these plugins? Or just don't do it on the plugins and instead on the libraries the plugins reference?

Reply:
this is correct.
we unit test the libs  and keep only the "glue" code in the recipe where applicable (it is a best practice that I admit not being reinforced yet)
(optional) we create an entire test project and run builds via DSS API and custom checks